src/terminal/input/terminalInput.cpp

```diff
@@ -513,209 +172,466 @@ TerminalInput::OutputType TerminalInput::HandleKey(const INPUT_RECORD& event)
     {
         // Note that we must return an empty string here to imply that we've handled
         // the event, otherwise the key press can still end up being submitted.
-        return MakeOutput({});
+        return _makeNoOutput();
     }
-    _lastVirtualKeyCode = keyEvent.wVirtualKeyCode;
+    _lastVirtualKeyCode = virtualKeyCode;
 
-    // The VK_BACK key depends on the state of Backarrow Key mode (DECBKM).
-    // If the mode is set, we should send BS. If reset, we should send DEL.
-    if (keyEvent.wVirtualKeyCode == VK_BACK)
+    // If this is a modifier, it won't produce output, so we can return early.
+    if (virtualKeyCode >= VK_SHIFT && virtualKeyCode <= VK_MENU)
     {
-        // The Ctrl modifier reverses the interpretation of DECBKM.
-        const auto backarrowMode = _inputMode.test(Mode::BackarrowKey) != WI_IsAnyFlagSet(keyEvent.dwControlKeyState, CTRL_PRESSED);
-        const auto seq = backarrowMode ? L'\x08' : L'\x7f';
-        // The Alt modifier adds an escape prefix.
-        if (WI_IsAnyFlagSet(keyEvent.dwControlKeyState, ALT_PRESSED))
-        {
-            return _makeEscapedOutput(seq);
-        }
-        else
-        {
-            return MakeOutput({ &seq, 1 });
-        }
+        return _makeNoOutput();
     }
 
-    // When the Line Feed mode is set, a VK_RETURN key should send both CR and LF.
-    // When reset, we fall through to the default behavior, which is to send just
-    // CR, or when the Ctrl modifier is pressed, just LF.
-    if (keyEvent.wVirtualKeyCode == VK_RETURN && _inputMode.test(Mode::LineFeed))
+    // Keyboards that have an AltGr key will generate both a RightAlt key press
+    // and a fake LeftCtrl key press. In order to support key combinations where
+    // the Ctrl key is manually pressed in addition to the AltGr key, we have to
+    // be able to detect when the Ctrl key isn't genuine. We do so by tracking
+    // the time between the Alt and Ctrl key presses, and only consider the Ctrl
+    // key to really be pressed if the difference is more than 50ms.

```

Comment:
okay this is pretty clever I'll give you that

Reply:
I can't entirely take credit for this. I did some research to see how other people dealt with the issue, and the typical approach was to handle it at the time the <kbd>Ctrl</kbd> key message was read from the queue. The idea was that you would peek the next message looking for an <kbd>Alt</kbd> key, and compare the two `MSG.time` fields. If it was a fake <kbd>Ctrl</kbd>+<kbd>Alt</kbd> combo, they would have identical values, so you would know to ignore the <kbd>Ctrl</kbd>.

But that exact approach wouldn't have worked for us over conpty, so that's why I track the time manually and allow a bit of leeway in the time difference, but the general idea is the same.